### PR TITLE
[WIP] Explore the use of IoUring.Concurrent (#51)

### PR DESCRIPTION
**Pull request by tkp1n**
Opening this as a draft to get early feedback not only in terms of devs looking into it but also in terms of running benchmarks.

#### Progress so far:
- [x] Use `IoUring.Concurrent.ConcurrentRing` over `IoUring.Ring` 
- [ ] Let threads submit directly to the submission queue
- [ ] Figure out how to best react if the ring is full and a thread wants to add a submission
- [ ] Benchmark and compare against master to see if any of this moves the needle (in the right direction)

@tmds or @adamsitnik It would be great to get some preliminary numbers now to see how far using `ConcurrentRing` over `Ring` sets us back as a starting point for optimizations... 

#### TODOs that could be solved along the way:

https://github.com/tmds/Tmds.LinuxAsync/blob/601c88372d14ef5b784d574ec741d3fb5df6d666/src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringThread.cs#L136

https://github.com/tmds/Tmds.LinuxAsync/blob/601c88372d14ef5b784d574ec741d3fb5df6d666/src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs#L88-L91

#### Related upstream work in [tkp1n/IoUring](https://github.com/tkp1n/IoUring) to be done before this could be merged (in no particular order):
- [x] Enable `IoUring.Concurrent` to be used stand-alone (without PackageReference to `IoUring`) tkp1n/IoUring#32
- [x] Polish API of `ConcurrentRing` as it is much more verbose than that of `Ring` tkp1n/IoUring#33
- [ ] Investigate a lock-free option for `SubmitAndWait` tkp1n/IoUring#34

_Note:_ I try to keep this description up to date as the work on this PR evolves.

### CODE REVIEW

**Comment by tmds** [comment]
>  Use IoUring.Concurrent.ConcurrentRing over IoUring.Ring

I prefer there to be a top-level option to choose between these implementation. You can duplicate code/add abstractions to achieve this.

**Comment by tkp1n** [comment]
Makes sense. Should simplify comparison later on as well...

File: src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs [hunk 1]
```diff
@@ -232,7 +247,7 @@ public unsafe void SubmitAndWait(bool mayWait)
                 bool waitForCompletion = !operationsRemaining && mayWait;
 
                 // io_uring_enter
-                SubmitResult result = _ring!.SubmitAndWait(minComplete: waitForCompletion ? 1U : 0, out _);
+                var result = _ring!.SubmitAndWait(minComplete: waitForCompletion ? 1U : 0, out _);
```
Comment (tmds) [inline src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs:250]: Currently we never wait for a completions when there are submissions remaining.
So if the eventfd read was not yet submitted, we can never be waiting.

Is it possible we now wait for completions without being sure the eventfd read is submitted?
That would mean we may not able to unblock the wait by writing to the eventfd.

File: src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs [hunk 2]
```diff
@@ -156,10 +157,11 @@ private void AddNewOperation(ulong key, Operation operation)
 
             private unsafe bool WriteSubmissions()
             {
-                Ring ring = _ring!;
+                ConcurrentRing ring = _ring!;
                 int iovIndex = _iovsUsed;
                 int sqesAvailable = ring.SubmissionEntriesAvailable;
```
Comment (tkp1n) [inline src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs:162]: `ring.SubmissionEntriesAvailable` returns a snapshot that could change (increase) while iterating below. We need to find an elegant way to continue submitting, while there are both new operations and enough (1 or 2 depending on Op) SQEs available.

File: src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs (hunk 1 above)
Comment (tkp1n) [reply src/Tmds.LinuxAsync/IOUringAsyncEngine.IOUringExecutionQueue.cs:250]: I think that could happen, yes. I have to put some more thought into it. Off the top of my hread we could:
* somehow track the submission of the eventfd read?
* only add the eventfd read as (likely) only SQE if there are no other SQEs prepared by the time of calling SubmitAndWait.

**Comment by tmds** [comment]
@tkp1n will you be pursuing this further?

**Comment by antonfirsov** [comment]
I'm skeptical whether it's worth to continue with this:
- Current bottlenecks are related to OS and ThreadPool behavior, I expect the gains to be relatively small.
- For productization, we will use liburing anyways. It's better to experiment with solutions which are available there.

**Comment by tkp1n** [comment]
> @tkp1n will you be pursuing this further?

Currently I think there are lower hanging fruits to focus on... so no, not at the moment. I'll close this PR for now.

> Current bottlenecks are related to OS and ThreadPool behavior, I expect the gains to be relatively small.

I generally agree, but small gains are gains nevertheless. :wink: 

> For productization, we will use liburing anyways. It's better to experiment with solutions which are available there.

I understand the approach, be aware though that using liburing will introduce a lot of additional interop calls (one per `sqe` preparation = one or two per operation) compared to the managed implementation only _interoping_ once per syscall...

**Pull request closed without merging**